[RLlib; docs] Docs do-over (new API stack): `ConnectorV2` documentation (part I). (#53732)
ray-project/ray: doc/source/rllib/connector-v2.rst

**Connector pipelines** are the components that translate this episode data into tensor batches
readable by neural network models right before the model forward pass.

.. figure:: images/connector_v2/generic_connector_pipeline.svg
    :width: 1000
    :align: left

    **Generic ConnectorV2 Pipeline**: All pipelines consist of one or more :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` pieces.
    When calling the pipeline, you pass in a list of Episodes, the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` instance,
    and a batch, which initially might be an empty dict.
    Each :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` piece in the pipeline takes its predecessor's output,
    starting on the left side with the batch, performs some transformations on the episodes, the batch, or both, and passes everything
    on to the next piece. Thereby, all :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` pieces can read from and write to the
    provided episodes, add any data from these episodes to the batch, or change the data that's already in the batch.
    The pipeline then returns the output batch of the last piece.

> Note that the batch output of the pipeline lives only as long as the succeeding

Author: done

Also, the Learner pipeline operates on the same episodes that have already passed through both env-to-module and module-to-env pipelines
and thus might have undergone changes.

Three ConnectorV2 pipeline types
--------------------------------

There are three different types of connector pipelines in RLlib:

1) :ref:`Env-to-module pipeline <env-to-module-pipeline-docs>`, which creates tensor batches for action computing forward passes.
2) Module-to-env pipeline, which translates a model's output into RL environment actions.
3) Learner connector pipeline, which creates the train batch for a model update.

The :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` API is an extremely powerful tool for
customizing your RLlib experiments and algorithms. It allows you to take full control over accessing, changing, and re-assembling
the episode data collected from your RL environments or your offline RL input files as well as controlling the exact
nature and shape of the tensor batches that RLlib feeds into your models for computing actions or losses.

.. figure:: images/connector_v2/location_of_connector_pipelines_in_rllib.svg
    :width: 900
    :align: left

>     **ConnectorV2 Pipelines**: Connector pipelines convert episodes into batched data, which a neural network can process

Author: simplified the block a bit

    (env-to-module and Learner) or convert your model's output into actions, which your RL environment needs for stepping (module-to-env).
    The env-to-module pipeline, located on an :py:class:`~ray.rllib.env.env_runner.EnvRunner`, takes a list of
    episodes as input and outputs a batch for an :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` forward pass
    that computes the next action. The module-to-env pipeline on the same :py:class:`~ray.rllib.env.env_runner.EnvRunner`
    takes the output of that :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` and converts it into actions
    for the next call to your RL environment's `step()` method.
    Lastly, a Learner connector pipeline, located on a :py:class:`~ray.rllib.core.learner.learner.Learner`
    worker, converts a list of episodes into a train batch for the next :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` update.

> 

Author: added "possibly vectorized" to text

            # Convert an input observation (int) into a one-hot (float) tensor.
            # Note that 99% of all connectors in RLlib operator in the "numpy space".
            new_obs = np.zeros(shape=self.observation_space.shape, dtype=np.float32)
            new_obs[observation] = 1.0
            return new_obs

Now that the preprocessor is written, you can plug it into your :py:class:`~ray.rllib.algorithms.algorithm_config.AlgorithmConfig`
object and run an experiment with a Discrete observation env (observations are integers), for
example `FrozenLake-v1 <https://gymnasium.farama.org/environments/toy_text/frozen_lake/>`__.

.. testcode::

    from ray.rllib.algorithms.ppo import PPOConfig

    config = (
        PPOConfig()

        # Configure a simple 2x2 grid-world.
        # ____
        # |S |  <- S=start position
        # | G|  <- G=goal position
        # ----
        .environment("FrozenLake-v1", env_config={"desc": ["SF", "FG"]})

        # Plug our custom connector piece into the env-to-module pipeline.
        .env_runners(env_to_module_connector=lambda env: OneHot())
    )
    algo = config.build()
    # Train one iteration.
    print(algo.train())

> ConnectorV2 pipelines in more detail

Reviewer: Question:
- Is the following detail necessary to work with the pipelines described above?

If yes, shouldn't this info precede the general descriptions to lay the foundation?

------------------------------------

Before diving into more complex and more powerful customization options of the different
pipeline types (and walking through more examples), the generic inner workings of a

> ConnectorV2 pipeline should be explained in more detail.

Reviewer: Passive voice. Just tell me what the section describes, don't tell me I need to learn this before learning more. Organize the content so that it makes sense in context and/or provide users with sufficient prompts and entry points to discover the content relevant to them.

.. figure:: images/connector_v2/generic_connector_pipeline.svg
    :width: 1000
    :align: left

    **Generic ConnectorV2 Pipeline**: All pipelines consist of one or more :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` pieces.

>     When the pipeline is called (for example by an :py:class:`~ray.rllib.env.env_runner.EnvRunner` or a :py:class:`~ray.rllib.core.learner.learner.Learner`),

Reviewer: "is called" passive

    a list of Episodes and the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` instance are provided.
    Each :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` in the pipeline takes its predecessor's output
    (starting on the left side with an initial, empty batch), performs some transformations on the data and passes it on to the next
    piece in the pipeline. Thereby, all :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` pieces can read from and write into the
    provided episodes, add any data from these episodes to the batch, or change the data already in the batch.
    The pipeline then returns the last piece's batch output (right side).

> Very importantly, note that the batch output of the pipeline always only lives as long as

Reviewer: ```suggestion
Note that the batch output of the pipeline always only lives as long as
```

"note" already indicates something is important

> the following :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule` forward pass and is discarded right after that.

Reviewer: "is discarded" passive

However, the list of episodes might survive longer. For example
if a env-to-module pipeline reads an observation from one episode, changes that observation, and then writes it
back into the episode, the subsequent module-to-env pipeline is able to see this change.
The same is true for the Learner pipeline: The episodes that it operates on
have already been through both env-to-module and module-to-env pipelines and might have been changed by these.

> Writing more complex custom ConnectorV2s

Reviewer: I think this is a new page on "ConnectorV2 examples" or similar title.

In general, when pages get too long it's good to break into sub-pages. This section now has a different intent than the preceding.

You were writing a conceptual overview "What are ConnectorV2s?". Now you're writing a guide on "Achieve X outcome with ConnectorV2" using examples.

----------------------------------------

> Besides simply transforming one observation at a time in a preprocessor-style setup as shown preceding,

Reviewer: Don't use "preceding".

Instead, "The following sections show examples for customizing ConnectorV2 pipelines. For a simple example, see :ref:(link back to that section)."

> here are four more examples showcasing more complex customizations of the different :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` pipelines:

Reviewer: You say 4 but I think I only saw like 2? Maybe I miscounted.

> * How to add rewards and or actions to the batch.

Reviewer: You say 4 above, but here you list 3 things?

This is purely navigational, which means there should be links to each example from this location. You can just use the section header instead of rewriting.

* How to stack the N most recent observations.
* How to manipulate the action logits coming from the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`.

> Adding past rewards to the Model's input

Reviewer: ```suggestion
Add past rewards to the model's input
```

Not clear to me why model should be capitalized, but I might not have the RLlib context.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If you want to add the most recent reward to be part of the next observation going into your

Reviewer: ```suggestion
To add the most recent reward to the next observation going into your
```

They're reading this because they want to do it, just tell them how to do it.

> :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`, you can write a custom

Reviewer: ```suggestion
:py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`, write a custom
```

It's not that they can do something, this is what they must do to achieve the aim expressed.

> env-to-module :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` piece like so:

Reviewer: ```suggestion
env-to-module :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` piece as in the following example:
```

Both "like" and "so" are hard for machine translation and English language learners. Avoid always.

            # For simplicity, assert input obs space is a 1D Box
            assert isinstance(in_obs_space, gym.spaces.Box) and len(in_obs_space.shape) == 1
            return gym.spaces.Box(
                low=list(in_obs_space.low) + [float("-inf")],
                high=list(in_obs_space.high) + [float("inf")],
                shape=(in_obs_space.shape[0] + 1,),
                dtype=in_obs_space.dtype,
            )

        # Define the actual transformation.
        def __call__(self, *, rl_module, batch, episodes, **kwargs):
            # Loop through all `episodes`.
            for single_agent_episode in self.single_agent_episode_iterator(episodes):
                # Get last reward (or 0.0 if right after `env.reset`).
                reward = single_agent_episode.get_reward(-1, fill=0.0)
                # Get last observation.
                obs = single_agent_episode.get_observation(-1)
                # Append reward to obs array.
                new_obs = np.append(obs, reward)

                # Write new observation back into the episode.
                single_agent_episode.set_observation(new_value=new_obs, at_index=-1)

            return batch

.. tip::

>     There is already an off-the-shelf ConnectorV2 piece available for users, which performs the task of

Reviewer: Just checking: are you saying that the above example implements this from scratch instead of using the built-in? Why not just use the built-in and reduce the code complexity for your example?

    adding the `N` most recent rewards and/or `M` most recent actions to the observations:

    .. code-block:: python

        from ray.rllib.connectors.env_to_module.prev_actions_prev_rewards import PrevActionsPrevRewards

        config.env_runners(
            env_to_module_connector=lambda env: PrevActionsPrevRewards(n_prev_rewards=N, n_prev_actions=M),
        )

If you plug in this custom :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` class into your algorithm config
(`config.env_runners(env_to_module_connector=lambda env: AddLastRewardToObservations())`),
your model should receive observations that have the most recent reward "attached" to it.

> Notice that in the example here, the transformed observations were written right back into the given episodes

Reviewer: ```suggestion
In the provided example, the transformed observations were written right back into the given episodes
```

Not sure the actor here in your passive voice "were written", but rewrite that section so that "X writes transformed observations into the given episodes instead of placing them into the batch."

> and not placed into the `batch`. This strategy of writing back those data that was pulled from episodes right back

Reviewer: "was pulled" is passive.

into the same episodes makes sure that from this point on, only the changed data is visible to all subsequent components (for example
other ConnectorV2 pieces in the same pipeline or other ConnectorV2 pipelines). It does not touch the `batch`.

> However, knowing that one of the subsequent :ref:`default env-to-module pieces <default-env-to-module-pipeline>`

Reviewer: You lost me here. What is this caveat aside all about?

> is going to do exactly that, we can defer this task (of populating the batch with our changed data) to these default pieces.

Reviewer: Who is the "we" here? The programmer? The RLlib engine?

This is why you should never use "we" or "our" in docs.

The next example, however, demonstrates a specific case - observation stacking - where this strategy fails and in
which users should instead manipulate the `batch` directly.

Stacking the N most recent observations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> If users would like to write a custom env-to-module connector that stacks the `N` most recent observations and feeds

Reviewer: If you need to...

> this stack of observations into their RLModule (for example in an attention/transformer architecture), they can

Reviewer: Instead of "they can achieve this, too" just say "You must subclass and override..."

achieve this, too, by subclassing `ConnectorV2` and overriding the `__call__` method.

> However, in this case, the implementation shouldn't write back the stacked observations into the episode

Reviewer: Rewrite this as a directive.

"Do not write back stacked observations to the episode. Manipulate the batch directly"

(as updated observation), because doing so would make the next call to the same ConnectorV2 piece to look back onto

> an already stacked previous observation. Instead, users should manipulate the `batch` directly, as in this example:

Reviewer: "As in the following example:"

            for single_agent_episode in self.single_agent_episode_iterator(episodes):
                # Get the 10 most recent observations from the episodes.
                last_10_obs = single_agent_episode.get_observations(
                    indices=[-10, -9, -8, -7, -6, -5, -4, -3, -2, -1], fill=0.0
                )
                # Concatenate the two observations.
                new_obs = np.concatenate(last_10_obs, axis=0)

                # Add the new observation to the `batch` using the
                # `ConnectorV2.add_batch_item()` utility.
                self.add_batch_item(
                    batch=batch,
                    column=Columns.OBS,
                    item_to_add=new_obs,
                    single_agent_episode=single_agent_episode,
                )

                # Note that we do not write the stacked observations back into the episode
                # as this would interfere with the next call of this same connector (it
                # would try to stack already stacked observations and thus produce a shape error).

            # Return batch (with stacked observations).
            return batch

> Now since the returned `batch` in the above env-to-module piece is discarded after the model forward pass

Reviewer: Is this meant as motivation for the next section of the docs, or is this describing something about this example? I don't understand the intent of this.

(and not stored in the episodes), we have to make sure to perform the framestacking again on the Learner
side of things.

> .. tip::

Reviewer: Again, is this tip something that could simplify the above code? Show the simplest code in examples.

    There are already off-the-shelf ConnectorV2 pieces available to users. These perform the task of
    stacking the last `N` observations in both the env-to-module and Learner pipelines:

    .. code-block:: python

        from ray.rllib.connectors.common.frame_stacking import FrameStacking

        # Framestacking on the EnvRunner side.
        config.env_runners(
            env_to_module_connector=lambda env: FrameStacking(num_frames=N),
        )
        # Then again on the Learner side.
        config.training(
            learner_connector=lambda obs_space, act_space: FrameStacking(num_frames=N, as_learner_conector=True),
        )

> Point to more examples in the `examples/connectors` folder.

Reviewer: This is a note to self it appears.

If we already have advanced examples in a folder, these examples should also be there, yes?
